lang/src/accounts/option.rs

```diff
@@ -0,0 +1,77 @@
+//! Option<T> type for optional accounts.
+//!
+//! # Example
+//! ```ignore
+//! #[derive(Accounts)]
+//! pub struct Example {
+//!     pub my_acc: Option<Account<'info, MyData>>
+//! }
+//! ```
+
+use std::collections::{BTreeMap, BTreeSet};
+
+use solana_program::account_info::AccountInfo;
+use solana_program::instruction::AccountMeta;
+use solana_program::pubkey::Pubkey;
+
+use crate::{Accounts, AccountsClose, AccountsExit, Result, ToAccountInfos, ToAccountMetas};
+
+impl<'info, T: Accounts<'info>> Accounts<'info> for Option<T> {
+    fn try_accounts(
+        program_id: &Pubkey,
+        accounts: &mut &[AccountInfo<'info>],
+        ix_data: &[u8],
+        bumps: &mut BTreeMap<String, u8>,
+        reallocs: &mut BTreeSet<Pubkey>,
+    ) -> Result<Self> {
+        // There are three cases we have to handle. We don't care if
+        // accounts is empty, so if that's the case we return None. This
+        // allows adding optional accounts at the end of the Accounts
+        // struct without causing a breaking change. This is safe and will error
+        // out if a required account is then added after the optional account and
+        // the accounts aren't passed in.
+        if accounts.is_empty() {
```

Comment:
Allowing the instructions to be simply missing an account idx, even if optional, feels a little messy possibly. Particularly when it comes to clients that are putting together instructions. They should probably be consistently padding out all of the account idxs for all the accounts, not sometimes forgetting to add in the extra idxs. 

I focus on the idx aspect here, because padding out the entire instruction doesn't take a bunch of extra bytes, it's just 1 byte (i think?) per account index, since the program id is already in the list of accounts.

Reply:
Essentially `if accounts.is_empty() ` should return an error, anything else is inconsistent with how all the other accounts types work. 

While there might be outlying reasons to do this, it just adds to the complexity of the account parsing system.

Reply:
Are there any use cases or benefits of allowing ommiting pubkeys for optional accounts? I agree with henry that it would be better to not introduce the inconsistency unless there's a concrete use case.

Reply:
Biased since I work primarily on open source protocol programs, but if that weren't a think we'd be forced to use remaining accounts again.